File: shorturl-app/src/main/java/pl/mpakula/shorturl/application/ApiErrorController.java
package pl.mpakula.shorturl.application;

import lombok.RequiredArgsConstructor;
import org.springframework.boot.web.servlet.error.ErrorController;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;

import javax.servlet.http.HttpServletRequest;
import java.util.Map;
import java.util.Optional;

import static javax.servlet.RequestDispatcher.ERROR_STATUS_CODE;

@Controller
@RequiredArgsConstructor
class ApiErrorController implements ErrorController {

    private final Map<HttpStatus, String> errorPages;

    @Override
    public String getErrorPath() {
        throw new UnsupportedOperationException();
    }

    @RequestMapping("/error")
    String handleError(HttpServletRequest request) {
        return Optional.ofNullable(request)
                .map(r -> (Integer) r.getAttribute(ERROR_STATUS_CODE))
                .map(HttpStatus::valueOf)
                .map(errorPages::get)
                .orElse("error");
    }

}

File: application/src/main/kotlin/pl/mpakula/shorturl/application/endpoints/ApiErrorController.kt
package pl.mpakula.shorturl.application.endpoints

import org.springframework.boot.web.servlet.error.ErrorController
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import javax.servlet.RequestDispatcher
import javax.servlet.http.HttpServletRequest

@Controller
internal class ApiErrorController(private val errorPages: Map<HttpStatus, String>) : ErrorController {

    override fun getErrorPath(): String {
        throw UnsupportedOperationException()
    }

    @RequestMapping("/error")
    fun handleError(request: HttpServletRequest): String {
        return request
            .let { it.getAttribute(RequestDispatcher.ERROR_STATUS_CODE) as Int? }
            ?.let(HttpStatus::valueOf)
            ?.let { errorPages[it] }
            ?: ("error")
    }
}